crates/wysiwyg/src/dom/dom_list_methods.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright 2022 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Methods on Dom that modify its contents and are guaranteed to conform to
+//! our invariants e.g. no empty text nodes, no adjacent text nodes.
+
+use core::panic;
+
+use crate::{DomHandle, DomNode, ListType, UnicodeString};
+
+use super::nodes::ContainerNode;
+use super::Dom;
+
+impl<S> Dom<S>
+where
+    S: UnicodeString,
+{
+    pub fn wrap_nodes_in_list(
+        &mut self,
+        list_type: ListType,
+        handles: Vec<&DomHandle>,
+    ) {
+        let first_handle = handles[0];
+        let mut removed_nodes = Vec::new();
+        for handle in handles.iter().rev() {
+            removed_nodes.push(self.remove(handle));
+        }
+        removed_nodes.reverse();
+
+        let mut list_item = ContainerNode::new_list_item(removed_nodes);
+        // Set an arbitrary handle allows us to transform this while detached from DOM.
+        list_item.set_handle(DomHandle::root());
+        let mut list_items = Vec::new();
+        let mut line_break_positions = list_item.line_break_positions();
+
+        while let Some(position) = line_break_positions.pop() {
+            let mut sliced = list_item.slice_after(position);
```

Comment:
Some comment here explaining how we're splitting the nodes by line breaks would be nice.